### ISSUE TITLE
Dell S4148F-ON (OS10) Error polling ospf module

### ISSUE THREAD

**Issue by zippanto** [librenms/librenms#14000]
### The problem

We get an "Error polling ospf module" error in the event log when polling Dell S4148F-ON switches running OS10 10.5.1.4

We have come across the issue #13592 and have asked Dell to provide us the fix. However this fix didn't see to make any difference for us.

Please see below for relevant parts of the poller output:
```
bash-5.1# /opt/librenms/poller.php -h dell-switch -m ospf -v -d
...
#### Load poller module ospf ####
...
4 Neighbours: SNMP['/usr/bin/snmpbulkwalk' '-M' '/opt/librenms/mibs:/opt/librenms/mibs/dell' '-v2c' '-c' 'public' '-OQXUt' '-Os' 'udp:dell-switch:161' 'OSPF-MIB::ospfNbrTable']
ospfNbrTable = No Such Object available on this agent at this OID

SNMP query failed. Exit Code: 0 Empty: false Bad String: No Such Object available on this agent at this OID
SQL[delete from `ospf_nbrs` where `ospf_nbrs`.`device_id` = ? and `ospf_nbrs`.`device_id` is not null and `context_name` = ? and 1 = 1 [85,""] 12.7ms]

0 TOS Metrics: SNMP['/usr/bin/snmpbulkwalk' '-M' '/opt/librenms/mibs:/opt/librenms/mibs/dell' '-v2c' '-c' 'public' '-OQXUt' '-Os' 'udp:dell-switch:161' 'OSPF-MIB::ospfIfMetricTable']
ospfIfMetricIpAddress[4.1.2.3][4][0].0 = 1.2.3.4
ospfIfMetricIpAddress[4.5.6.7][60][0].0 = 5.6.7.60
ospfIfMetricIpAddress[4.9.10.11][1][0].0 = 9.10.11.1
ospfIfMetricAddressLessIf[4.1.2.3][4][0].0 = 0
ospfIfMetricAddressLessIf[4.5.6.7][60][0].0 = 0
ospfIfMetricAddressLessIf[4.9.10.11][1][0].0 = 0
ospfIfMetricTOS[4.1.2.3][4][0].0 = 0
ospfIfMetricTOS[4.5.6.7][60][0].0 = 0
ospfIfMetricTOS[4.9.10.11][1][0].0 = 0
ospfIfMetricValue[4.1.2.3][4][0].0 = 10
ospfIfMetricValue[4.5.6.7][60][0].0 = 10
ospfIfMetricValue[4.9.10.11][1][0].0 = 10
ospfIfMetricStatus[4.1.2.3][4][0].0 = active
ospfIfMetricStatus[4.5.6.7][60][0].0 = active
ospfIfMetricStatus[4.9.10.11][1][0].0 = active

SQL[select `ospf_port_id` from `ospf_ports` where `ospfIfIpAddress` = ? and `context_name` = ? limit 1 ["4.1.2.3",""] 0.68ms]

SQL[select * from `ospf_ports` where (`device_id` = ? and `ospf_port_id` is null and `context_name` = ?) limit 1 [85,""] 0.53ms]

Error polling ospf module for dell-switch. PDOException: SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'ospf_port_id' cannot be null in /opt/librenms/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:117
Stack trace:
#0 /opt/librenms/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php(117): PDOStatement->execute()
#1 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(501): Doctrine\DBAL\Driver\PDOStatement->execute()
#2 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(705): Illuminate\Database\Connection->Illuminate\Database\{closure}()
#3 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(672): Illuminate\Database\Connection->runQueryCallback()
#4 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(502): Illuminate\Database\Connection->run()
#5 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(454): Illuminate\Database\Connection->statement()
#6 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Query/Processors/Processor.php(32): Illuminate\Database\Connection->insert()
#7 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2999): Illuminate\Database\Query\Processors\Processor->processInsertGetId()
#8 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(1655): Illuminate\Database\Query\Builder->insertGetId()
#9 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1188): Illuminate\Database\Eloquent\Builder->__call()
#10 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1153): Illuminate\Database\Eloquent\Model->insertAndSetId()
#11 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(994): Illuminate\Database\Eloquent\Model->performInsert()
#12 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(509): Illuminate\Database\Eloquent\Model->save()
#13 /opt/librenms/vendor/laravel/framework/src/Illuminate/Support/helpers.php(263): Illuminate\Database\Eloquent\Builder->Illuminate\Database\Eloquent\{closure}()
#14 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(510): tap()
#15 /opt/librenms/vendor/laravel/framework/src/Illuminate/Support/Traits/ForwardsCalls.php(23): Illuminate\Database\Eloquent\Builder->updateOrCreate()
#16 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(2132): Illuminate\Database\Eloquent\Model->forwardCallTo()
#17 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(2144): Illuminate\Database\Eloquent\Model->__call()
#18 /opt/librenms/LibreNMS/Modules/Ospf.php(184): Illuminate\Database\Eloquent\Model::__callStatic()
#19 [internal function]: LibreNMS\Modules\Ospf->LibreNMS\Modules\{closure}()
#20 /opt/librenms/LibreNMS/Data/Source/SnmpResponse.php(211): call_user_func()
#21 [internal function]: LibreNMS\Data\Source\SnmpResponse->LibreNMS\Data\Source\{closure}()
#22 /opt/librenms/vendor/laravel/framework/src/Illuminate/Collections/Collection.php(677): array_map()
#23 /opt/librenms/LibreNMS/Data/Source/SnmpResponse.php(212): Illuminate\Support\Collection->map()
#24 /opt/librenms/LibreNMS/Modules/Ospf.php(188): LibreNMS\Data\Source\SnmpResponse->mapTable()
#25 /opt/librenms/includes/polling/ospf.inc.php(8): LibreNMS\Modules\Ospf->poll()
#26 /opt/librenms/includes/polling/functions.inc.php(329): include('/opt/librenms/i...')
#27 /opt/librenms/poller.php(138): poll_device()
#28 {main}

Next Doctrine\DBAL\Driver\PDO\Exception: SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'ospf_port_id' cannot be null in /opt/librenms/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDO/Exception.php:18
Stack trace:
#0 /opt/librenms/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php(119): Doctrine\DBAL\Driver\PDO\Exception::new()
#1 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(501): Doctrine\DBAL\Driver\PDOStatement->execute()
#2 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(705): Illuminate\Database\Connection->Illuminate\Database\{closure}()
#3 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(672): Illuminate\Database\Connection->runQueryCallback()
#4 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(502): Illuminate\Database\Connection->run()
#5 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(454): Illuminate\Database\Connection->statement()
#6 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Query/Processors/Processor.php(32): Illuminate\Database\Connection->insert()
#7 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2999): Illuminate\Database\Query\Processors\Processor->processInsertGetId()
#8 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(1655): Illuminate\Database\Query\Builder->insertGetId()
#9 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1188): Illuminate\Database\Eloquent\Builder->__call()
#10 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1153): Illuminate\Database\Eloquent\Model->insertAndSetId()
#11 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(994): Illuminate\Database\Eloquent\Model->performInsert()
#12 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(509): Illuminate\Database\Eloquent\Model->save()
#13 /opt/librenms/vendor/laravel/framework/src/Illuminate/Support/helpers.php(263): Illuminate\Database\Eloquent\Builder->Illuminate\Database\Eloquent\{closure}()
#14 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(510): tap()
#15 /opt/librenms/vendor/laravel/framework/src/Illuminate/Support/Traits/ForwardsCalls.php(23): Illuminate\Database\Eloquent\Builder->updateOrCreate()
#16 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(2132): Illuminate\Database\Eloquent\Model->forwardCallTo()
#17 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(2144): Illuminate\Database\Eloquent\Model->__call()
#18 /opt/librenms/LibreNMS/Modules/Ospf.php(184): Illuminate\Database\Eloquent\Model::__callStatic()
#19 [internal function]: LibreNMS\Modules\Ospf->LibreNMS\Modules\{closure}()
#20 /opt/librenms/LibreNMS/Data/Source/SnmpResponse.php(211): call_user_func()
#21 [internal function]: LibreNMS\Data\Source\SnmpResponse->LibreNMS\Data\Source\{closure}()
#22 /opt/librenms/vendor/laravel/framework/src/Illuminate/Collections/Collection.php(677): array_map()
#23 /opt/librenms/LibreNMS/Data/Source/SnmpResponse.php(212): Illuminate\Support\Collection->map()
#24 /opt/librenms/LibreNMS/Modules/Ospf.php(188): LibreNMS\Data\Source\SnmpResponse->mapTable()
#25 /opt/librenms/includes/polling/ospf.inc.php(8): LibreNMS\Modules\Ospf->poll()
#26 /opt/librenms/includes/polling/functions.inc.php(329): include('/opt/librenms/i...')
#27 /opt/librenms/poller.php(138): poll_device()
#28 {main}

Next Illuminate\Database\QueryException: SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'ospf_port_id' cannot be null (SQL: insert into `ospf_ports` (`device_id`, `ospf_port_id`, `context_name`, `ospfIfMetricIpAddress`, `ospfIfMetricAddressLessIf`, `ospfIfMetricTOS`, `ospfIfMetricValue`, `ospfIfMetricStatus`) values (85, ?, , 1.2.3.4, 0, 0, 10, active)) in /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php:712
Stack trace:
#0 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(672): Illuminate\Database\Connection->runQueryCallback()
#1 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(502): Illuminate\Database\Connection->run()
#2 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php(454): Illuminate\Database\Connection->statement()
#3 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Query/Processors/Processor.php(32): Illuminate\Database\Connection->insert()
#4 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php(2999): Illuminate\Database\Query\Processors\Processor->processInsertGetId()
#5 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(1655): Illuminate\Database\Query\Builder->insertGetId()
#6 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1188): Illuminate\Database\Eloquent\Builder->__call()
#7 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(1153): Illuminate\Database\Eloquent\Model->insertAndSetId()
#8 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(994): Illuminate\Database\Eloquent\Model->performInsert()
#9 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(509): Illuminate\Database\Eloquent\Model->save()
#10 /opt/librenms/vendor/laravel/framework/src/Illuminate/Support/helpers.php(263): Illuminate\Database\Eloquent\Builder->Illuminate\Database\Eloquent\{closure}()
#11 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php(510): tap()
#12 /opt/librenms/vendor/laravel/framework/src/Illuminate/Support/Traits/ForwardsCalls.php(23): Illuminate\Database\Eloquent\Builder->updateOrCreate()
#13 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(2132): Illuminate\Database\Eloquent\Model->forwardCallTo()
#14 /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php(2144): Illuminate\Database\Eloquent\Model->__call()
#15 /opt/librenms/LibreNMS/Modules/Ospf.php(184): Illuminate\Database\Eloquent\Model::__callStatic()
#16 [internal function]: LibreNMS\Modules\Ospf->LibreNMS\Modules\{closure}()
#17 /opt/librenms/LibreNMS/Data/Source/SnmpResponse.php(211): call_user_func()
#18 [internal function]: LibreNMS\Data\Source\SnmpResponse->LibreNMS\Data\Source\{closure}()
#19 /opt/librenms/vendor/laravel/framework/src/Illuminate/Collections/Collection.php(677): array_map()
#20 /opt/librenms/LibreNMS/Data/Source/SnmpResponse.php(212): Illuminate\Support\Collection->map()
#21 /opt/librenms/LibreNMS/Modules/Ospf.php(188): LibreNMS\Data\Source\SnmpResponse->mapTable()
#22 /opt/librenms/includes/polling/ospf.inc.php(8): LibreNMS\Modules\Ospf->poll()
#23 /opt/librenms/includes/polling/functions.inc.php(329): include('/opt/librenms/i...')
#24 /opt/librenms/poller.php(138): poll_device()
#25 {main}
SQL[insert into `eventlog` (`device_id`, `reference`, `type`, `datetime`, `severity`, `message`, `username`) values (?, ?, ?, ?, ?, ?, ?) [85,null,"poller","2022-05-30 12:49:28",5,"Error polling ospf module. Check log file for more details.",""] 1.84ms]
```

### Output of ./validate.php

```txt
/opt/librenms $ /opt/librenms/validate.php
====================================
Component | Version
--------- | -------
LibreNMS  | 22.5.0
DB Schema | 2022_04_08_085504_isis_adjacencies_table_add_index (236)
PHP       | 7.4.29
Python    | 3.9.7
MySQL     | 10.5.16-MariaDB-1:10.5.16+maria~focal
RRDTool   | 1.7.2
SNMP      | 5.9.1
====================================

[OK]    Installed from the official Docker image; no Composer required
[OK]    Database connection successful
[OK]    Database schema correct
[WARN]  IPv6 is disabled on your server, you will not be able to add IPv6 devices.
[WARN]  Updates are managed through the official Docker image
```

### Device info

Dell S4148F-ON

### Discovery output

```txt
available on request
```

### Poller output

```txt
available on request
```

**Comment by rujschafer** [comment]
@zippanto were you able to find a resolution to this issue? I was combing through my log for a different problem and noticed this. Running: 10.5.2.0P1 and afraid the answer is upgrade the OS.

**Comment by zippanto** [comment]
> @zippanto were you able to find a resolution to this issue? I was combing through my log for a different problem and noticed this. Running: 10.5.2.0P1 and afraid the answer is upgrade the OS.

We have not. I suspect the issue might be to do with using VRFs + OSPF. Perhaps an upgrade might fix it.

We did have a deeper look and it seemed like the format of the OSPF SNMP responses is different than what LibreNMS is expecting. But never pursued it any further.

**Comment by rujschafer** [comment]
I am working with Dell to do a software update this month. Ill let you know if it fixes the problem.

**Comment by rujschafer** [comment]
@zippanto just did the upgrade this morning and same thing. I think the error is in the OID response from SNMP. I think it is prepended with a "4" which makes the parsing not work.

**Comment by murrant** [comment]
Does this device not supply `ospfNbrTable`? Make sure you are not filtering that oid.

**Comment by zippanto** [comment]
@murrant The reason for no ospfNbrTable is that there are no neighbours at the time.

**Comment by rujschafer** [comment]
Any solution to this yet?

My next step is to contact Dell Support but I doubt that is going to yield what I want.

**Comment by rujschafer** [comment]
Thank you for the suggestion. I put it in but it didnt resolve the security constraint on the database. 

Listed here:
`Next Illuminate\Database\QueryException: SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'ospf_port_id' cannot be null (SQL: insert into `ospf_ports` (`device_id`, `ospf_port_id`, `context_name`, `ospfIfMetricIpAddress`, `ospfIfMetricAddressLessIf`, `ospfIfMetricTOS`, `ospfIfMetricValue`, `ospfIfMetricStatus`) values (23, ?, , www.xxx.yyy.zzz, 0, 0, 1, active)) in /opt/librenms/vendor/laravel/framework/src/Illuminate/Database/Connection.php:712
`

**Comment by murrant** [comment]
So, the security constraint isn't the problem, don't focus on that.

We need to get correct and parseable data from the device.

**Comment by rujschafer** [comment]
> So, the security constraint isn't the problem, don't focus on that.
> 
> We need to get correct and parseable data from the device.

sorry just saw this. What do you need?

**Comment by murrant** [comment]
@rujschafer run the poller with debug enabled (-d) and pastebin the output.

**Comment by rujschafer** [comment]
@murrant I was trying to post the pastebin but it looks like I hit the paste limit. Do you want just the OSPF module or are there other sections you want as well?

**Comment by laf** [comment]
Just the OSPF module should be fine.

**Comment by zippanto** [comment]
I'm going to close this as it's working for me now. I think it was fixed in db17036.
